useEffect generates additional (unwanted) error

**Scenario**
```js
class Root extends owl.Component {
    static template = owl.xml`<div />`;
    setup() {
        owl.useEffect(
            () => {
                throw new Error("Intentional error");
            },
            () => []
        ); 
    }
}

try {
    await owl.mount(Root, document.body, { dev: true });
} catch (err) {
    /** Error caught! */
}
```
**Problem**
Even though the error is caught, the execution of the `useEffect` hook was disturbed and its `cleanup` is not assigned, hence generating an additional error in the console:
```
owl.js:1941 Error: The following error occurred in onWillUnmount: "cleanup is not a function"
    at wrapError (owl.js:4839:23)
    at onWillUnmount (owl.js:4893:34)
    at Object.useEffect (owl.js:5450:9)
    at Root.setup (about:blank:6:13)
    at new ComponentNode (owl.js:2390:28)
    at App.makeNode (owl.js:5321:20)
    at App.mount (owl.js:5315:31)
    at Object.mount (owl.js:5359:35)
    at about:blank:16:15
    at about:blank:21:11
```
I think this error is unnecessary since it's a result of the fallback not being correctly applied, and the rendering process would have already been interupted by the first error anyway.

--- a/src/hooks.ts
+++ b/src/hooks.ts
@@ -57,7 +57,6 @@ export function useChildSubEnv(envExtension: Env) {
 // useEffect
 // -----------------------------------------------------------------------------
 
-const NO_OP = () => {};
 /**
  * @param {...any} dependencies the dependencies computed by computeDependencies
  * @returns {void|(()=>void)} a cleanup function that reverses the side
@@ -78,11 +77,11 @@ type Effect = (...dependencies: any[]) => void | (() => void);
  *      NaN !== NaN, which will cause the effect to rerun on every patch.
  */
 export function useEffect(effect: Effect, computeDependencies: () => any[] = () => [NaN]) {
-  let cleanup: () => void;
+  let cleanup: (() => void) | void;
   let dependencies: any[];
   onMounted(() => {
     dependencies = computeDependencies();
-    cleanup = effect(...dependencies) || NO_OP;
+    cleanup = effect(...dependencies);
   });
 
   onPatched(() => {
@@ -90,12 +89,14 @@ export function useEffect(effect: Effect, computeDependencies: () => any[] = () 
     const shouldReapply = newDeps.some((val, i) => val !== dependencies[i]);
     if (shouldReapply) {
       dependencies = newDeps;
-      cleanup();
-      cleanup = effect(...dependencies) || NO_OP;
+      if (cleanup) {
+        cleanup();
+      }
+      cleanup = effect(...dependencies);
     }
   });
 
-  onWillUnmount(() => cleanup());
+  onWillUnmount(() => cleanup && cleanup());
 }
 
 // -----------------------------------------------------------------------------

--- a/tests/components/hooks.test.ts
+++ b/tests/components/hooks.test.ts
@@ -632,5 +632,35 @@ describe("hooks", () => {
         "cleaning up for 1",
       ]);
     });
+
+    test("properly behaves when the effect function throws", async () => {
+      let originalconsoleError = console.error;
+      let originalconsoleWarn = console.warn;
+      console.error = jest.fn(() => {});
+      console.warn = jest.fn(() => {});
+      class MyComponent extends Component {
+        static template = xml`<div/>`;
+        setup() {
+          useEffect(
+            () => {
+              throw new Error("Intentional error");
+            },
+            () => []
+          );
+        }
+      }
+
+      try {
+        await mount(MyComponent, fixture);
+      } catch (e: any) {
+        expect(e.message).toBe("Intentional error");
+      }
+      // no console.error because the error has been caught in this test
+      expect(console.error).toHaveBeenCalledTimes(0);
+      console.error = originalconsoleError;
+      // 1 console.warn because app is destroyed
+      expect(console.warn).toHaveBeenCalledTimes(1);
+      console.warn = originalconsoleWarn;
+    });
   });
 });

